### PR TITLE
Clean up candidates

### PR DESCRIPTION
**Pull request by CoderZhi**
1. After hardfork, fix logic bug of updating candidate name/operator
2. Clean up candidate names and operator addresses

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/candidate_statemanager.go [hunk 1]
```diff
@@ -56,7 +56,7 @@ type (
 		Upsert(*Candidate) error
 		CreditBucketPool(*big.Int) error
 		DebitBucketPool(*big.Int, bool) error
-		Commit() error
+		Commit(bool) error
```
Comment (dustinxie) [inline action/protocol/staking/candidate_statemanager.go:59]: this fix should only happen one-time at the HF height, why need to add it as an extra input

File: action/protocol/staking/vote_reviser.go [hunk 2]
```diff
@@ -14,39 +14,74 @@ import (
...
+	if !ok {
+		return nil, false
+	}
+	return cands[1], true
 }
 
-func (vr *VoteReviser) storeToCache(height uint64, cands CandidateList) {
-	vr.cache[height] = cands
+func (vr *VoteReviser) storeToCache(height uint64, dirty, cands CandidateList) {
+	vr.cache[height] = [2]CandidateList{dirty, cands}
```
Comment (dustinxie) [inline action/protocol/staking/vote_reviser.go:84]: cache should only contain the correct candidates, we should not need to change the implementation here

File: action/protocol/staking/vote_reviser.go [hunk 3]
```diff
@@ -117,19 +157,21 @@ func (vr *VoteReviser) calculateVoteWeight(csm CandidateStateManager) (Candidate
...
-			return err
+		zap.Uint64("height", height), zap.Int("number of cands", len(cache[1])))
+	for i := 0; i < 2; i++ {
+		sort.Sort(cache[i])
+		for _, cand := range cache[i] {
+			if err := csm.Upsert(cand); err != nil {
+				return err
+			}
+			log.L().Info("committed revise action",
+				zap.String("name", cand.Name), zap.String("votes", cand.Votes.String()))
```
Comment (dustinxie) [inline action/protocol/staking/vote_reviser.go:173]: we should be able to do the fix w/o modifying these low-level implementations

File: action/protocol/staking/candidate_center.go [hunk 4]
```diff
@@ -440,14 +444,20 @@ func (cb *candBase) all() CandidateList {
 	return list
 }
 
-func (cb *candBase) commit(change *candChange) (int, error) {
+func (cb *candBase) commit(change *candChange, keepAliasBug bool) (int, error) {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:447]: if we could have the flag in the `candBase`, then I think it's inefficient to pass it down every time

File: action/protocol/staking/candidate_center.go [hunk 5]
```diff
@@ -17,7 +17,8 @@ import (
 type (
 	// candChange captures the change to candidates
 	candChange struct {
-		dirty map[string]*Candidate
+		candidates []*Candidate
+		dirty      map[string]*Candidate
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:21]: we should be able to do the fix w/o modifying these low-level implementations

File: action/protocol/staking/vote_reviser.go [hunk 6]
```diff
@@ -14,39 +14,74 @@ import (
...
 		if err != nil {
 			return err
 		}
-		vr.storeToCache(height, cands)
+		if vr.needCorrectCands(height) {
+			for _, c := range csm.DirtyView().candCenter.base.nameMap {
+				dirty = append(dirty, c)
+			}
+			for _, c := range csm.DirtyView().candCenter.base.operatorMap {
+				dirty = append(dirty, c)
```
Comment (dustinxie) [inline action/protocol/staking/vote_reviser.go:46]: reviser should not bother/deal with `candCenter.base.operatorMap`

File: action/protocol/staking/candidate_center.go (hunk 5 above)
Comment (Liuhaai) [reply action/protocol/staking/candidate_center.go:21]: could either `map` or `list` be used in the `candChange` struct?

File: action/protocol/staking/candidate_center.go [hunk 7]
```diff
@@ -138,7 +140,19 @@ func (m *CandidateCenter) SetDelta(l CandidateList) error {
...
+	if err != nil {
+		return err
+	}
+	m.size = size
+	m.change = nil
+	m.change = newCandChange()
+	return nil
+}
+
+// LegacyCommit writes the change into base with legacy logic
```
Comment (guo) [inline action/protocol/staking/candidate_center.go:153]: It is cleaner than the bool argument.

File: action/protocol/staking/vote_reviser.go [hunk 8]
```diff
@@ -35,14 +37,48 @@ func (vr *VoteReviser) Revise(csm CandidateStateManager, height uint64) error {
...
+	for _, c := range csm.DirtyView().candCenter.base.operatorMap {
+		retval = append(retval, c)
+	}
+	sort.Sort(retval)
+	ownerMap := map[string]*Candidate{}
+	for addr, owner := range csm.DirtyView().candCenter.base.ownerMap {
+		ownerMap[addr] = owner
+	}
+	for _, c := range cands {
+		if owner, ok := ownerMap[c.Owner.String()]; ok {
```
Comment (Liuhaai) [inline action/protocol/staking/vote_reviser.go:66]: if owner, ok := csm.DirtyView().candCenter.base.ownerMap[c.Owner.String()]; ok {}

**Review by Liuhaai — APPROVED** [review]

**Comment by Liuhaai** [comment]
Discussed offline. The idea of this pr is easier to understand

File: action/protocol/staking/vote_reviser.go [hunk 9]
```diff
@@ -35,14 +37,48 @@ func (vr *VoteReviser) Revise(csm CandidateStateManager, height uint64) error {
...
 }
 
+func (vr *VoteReviser) correctAliasCands(csm CandidateStateManager, cands CandidateList) (CandidateList, error) {
+	var retval CandidateList
+	for _, c := range csm.DirtyView().candCenter.base.nameMap {
+		retval = append(retval, c)
+	}
+	for _, c := range csm.DirtyView().candCenter.base.operatorMap {
+		retval = append(retval, c)
+	}
```
Comment (dustinxie) [inline action/protocol/staking/vote_reviser.go:59]: we should not use `name/operatorMap` here? don't quite understand the code.

here's my understanding of the intended logic per our discussion during v1.8.4 fix
input `cands` is the list of candidates (with vote correctly revised) in the statedb, as done in v1.8.4 fix, we saved the list of candidates that have been called `CandidateRegister/Update` actions during the affected period, these are candidates that can potentially have an alias in the candidate center
```
_, _, owners := readCandCenterStateFromStateDB()
```
so on top of `cands`, update the name/operator/reward of `owners`, that is what we need to do

Comment (dustinxie) [reply action/protocol/staking/vote_reviser.go:59]: another comment is that `VoteReviser` logic is clean, it just reads candidates/buckets from statedb, and tally votes again.

now it needs to deal with `csm.DirtyView().candCenter.xxx`, this adds unnecessary code coupling

Comment (CoderZhi) [reply action/protocol/staking/vote_reviser.go:59]: https://github.com/iotexproject/iotex-core/blob/e0578dacb9752611a4a94c4efe4774f7ece0f320/action/protocol/staking/protocol.go#L182 has called and loaded the data.
https://github.com/iotexproject/iotex-core/blob/ee04a051811a9c23b33a291fb66c2de957b78bf3/action/protocol/staking/handlers.go#L689 candcenter has been used in various places, which is a legacy issue

Comment (dustinxie) [reply action/protocol/staking/vote_reviser.go:59]: yes, ideally we should not use `csm.DirtyView().candCenter.xxx` in `protocol.go` and `handlers.go`
so we should not expand it into `vote_reviser.go` if possible

Comment (dustinxie) [reply action/protocol/staking/vote_reviser.go:59]: `name, operator, owners, err := readCandCenterStateFromStateDB(sr, height)`, will NOT be called at next HF height

**Review by raullenchai — APPROVED** [review]

**Pull request merged**